app/components/form/fields/TlsCertsField.tsx

```diff
@@ -0,0 +1,137 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, you can obtain one at https://mozilla.org/MPL/2.0/.
+ *
+ * Copyright Oxide Computer Company
+ */
+import { useState } from 'react'
+import type { Control } from 'react-hook-form'
+import { useController } from 'react-hook-form'
+
+import { Button, Error16Icon, FieldLabel, MiniTable, Modal } from '@oxide/ui'
+
+import { DescriptionField, FileField, NameField } from 'app/components/form'
+import type { SiloCreateInput } from 'app/forms/silo-create'
+import { useForm } from 'app/hooks'
+import { readBlobAsText } from 'app/util/file'
+
+export function TlsCertsField({ control }: { control: Control<SiloCreateInput> }) {
+  const [showAddCert, setShowAddCert] = useState(false)
+
+  const {
+    field: { value: items, onChange },
+  } = useController({ control, name: 'tlsCertificates' })
+
+  return (
+    <>
+      <div className="max-w-lg">
+        <FieldLabel id="tls-certificates-label" className="mb-3">
+          TLS Certificates
+        </FieldLabel>
+        {!!items.length && (
+          <MiniTable.Table className="mb-4">
+            <MiniTable.Header>
+              <MiniTable.HeadCell>Name</MiniTable.HeadCell>
+              {/* For remove button */}
+              <MiniTable.HeadCell className="w-12" />
+            </MiniTable.Header>
+            <MiniTable.Body>
+              {items.map((item, index) => (
+                <MiniTable.Row
+                  tabIndex={0}
+                  aria-rowindex={index + 1}
+                  aria-label={`Name: ${item.name}, Description: ${item.description}`}
+                  key={item.name}
+                >
+                  <MiniTable.Cell>{item.name}</MiniTable.Cell>
+                  <MiniTable.Cell>
+                    <button
+                      onClick={() => onChange(items.filter((i) => i.name !== item.name))}
+                    >
+                      <Error16Icon title={`remove ${item.name}`} />
+                    </button>
+                  </MiniTable.Cell>
+                </MiniTable.Row>
+              ))}
+            </MiniTable.Body>
+          </MiniTable.Table>
+        )}
+
+        <Button size="sm" onClick={() => setShowAddCert(true)}>
+          Add TLS certificate
+        </Button>
+      </div>
+
+      {showAddCert && (
+        <AddCertModal
+          onDismiss={() => setShowAddCert(false)}
+          onSubmit={async (values) => {
+            const keypair = {
+              cert: await readBlobAsText(values.cert),
```

Comment:
turns out you should be able to do `await values.cert.text()`

https://developer.mozilla.org/en-US/docs/Web/API/Blob/text

Reply:
Confirmed that works, that's fun. I also noticed while testing that if the file dropped on the input has newlines, `readBlobAsText` seems to only get the last line.